File: 19-SecondGreatLow/19.hs
import Data.List

--Uses arrays
secondGreatLow' :: [Int] -> String
secondGreatLow' arr = getResult $ getData
  where getData = sort . nub $ arr
        getResult [x] = show x ++ " " ++ show x
        getResult xs = (show $ xs !! 1) ++ " " ++ (show $ xs !! (length xs - 2))

--Avoids arrays. Folds through the list and keeps track of low and high numbers instead.
secondGreatLow :: [Int] -> String
secondGreatLow arr = (\((a,b),(c,d)) -> show b ++ " " ++ show d) results
  where results = foldr check ((max,max),(min,min)) arr
        max = maxBound :: Int
        min = minBound :: Int
        check x acc = checkGreat . checkLow $ acc
              where checkLow ((low,secLow),(great,secGreat))
                      | x < low      = ((x,low),(great,secGreat))
                      | x < secLow   = ((low,x),(great,secGreat))
                      | otherwise    = ((low,secLow),(great,secGreat))
                    checkGreat ((low,secLow),(great,secGreat))
                      | x > great    = ((low,secLow),(x,great))
                      | x > secGreat = ((low,secLow),(great,x))
                      | otherwise    = ((low,secLow),(great,secGreat))
main = do
  print $ secondGreatLow [32,10,5,23,1] == "5 23"
  print $ secondGreatLow [80,80]        == "80 80"
  print $ secondGreatLow [80,30]        == "80 30"

File: 19-SecondGreatLow/19.js
// Given an list of #'s, print the second highest and lowest numbers of it

function SecondGreatLow(arr) { 
  var data = arr.sort(function(a,b){
    return a > b;
  }).reduce(function(acc,x){
    return (x !== acc[acc.length -1]) ? acc.concat(x) : acc;
  },[]); //Reduce will remove duplicate numbers from the sorted list
  
  if(data.length > 1){
    return data[1] + " " + data[data.length - 2];
  }else{ 
    return data[0] + " " + data[0];
  }
}

console.log(SecondGreatLow([32,10,5,23,1]) === "5 23")
console.log(SecondGreatLow([80,80]) === "80 80")
console.log(SecondGreatLow([80,30]) === "80 30")
